packages/hardhat-core/test/internal/hardhat-network/provider/modules/hardhat.ts

```diff
@@ -365,6 +369,627 @@ describe("Hardhat module", function () {
+            DEFAULT_ACCOUNTS_ADDRESSES[0],
+            `0x${targetBalance.toString(16)}`,
+          ]);
+
+          // Assert: Ensure the new balance was set.
+          const newBalance = new BN(
+            (
+              await this.provider.send("eth_getBalance", [
+                DEFAULT_ACCOUNTS_ADDRESSES[0],
+              ])
+            ).replace("0x", ""),
+            "hex"
+          );
+          assert(targetBalance.eq(newBalance));
+        });
+
+        it("should not result in a modified state root", async function () {
+          // Arrange 1: Send a transaction, in order to ensure a pre-existing
+          // state root.
+          await this.provider.send("eth_sendTransaction", [
+            {
+              from: DEFAULT_ACCOUNTS_ADDRESSES[0],
+              to: DEFAULT_ACCOUNTS_ADDRESSES[1],
+              value: "0x100",
+            },
+          ]);
+
+          // Arrange 2: Capture the existing state root.
+          const oldStateRoot = (
+            await this.provider.send("eth_getBlockByNumber", ["latest", false])
+          ).stateRoot;
+
+          // Act: Set the new balance.
+          await this.provider.send("hardhat_setBalance", [
+            DEFAULT_ACCOUNTS_ADDRESSES[0],
+            intToHex(99),
+          ]);
+
+          // Assert: Ensure state root hasn't changed.
+          const newStateRoot = (
+            await this.provider.send("eth_getBlockByNumber", ["latest", false])
+          ).stateRoot;
+          assert.equal(newStateRoot, oldStateRoot);
+        });
+
+        it("should get changed balance by block", async function () {
+          // Arrange 1: Get current block number
+          const currentBlockNumber = await this.provider.send(
+            "eth_blockNumber"
+          );
+
+          // Arrange 2: Set a new balance
+          const targetBalance = new BN("123454321");
+          const targetBalanceHex = `0x${targetBalance.toString(16)}`;
+          await this.provider.send("hardhat_setBalance", [
+            DEFAULT_ACCOUNTS_ADDRESSES[0],
+            targetBalanceHex,
+          ]);
+
+          // Arrange 3: Mine a block
+          await this.provider.send("evm_mine");
+
+          // Act: Get the balance of the account in the previous block
+          const balancePreviousBlock = await this.provider.send(
+            "eth_getBalance",
+            [DEFAULT_ACCOUNTS_ADDRESSES[0], currentBlockNumber]
+          );
+
+          // Assert: Check that the balance is the one we set
+          assert.equal(balancePreviousBlock, targetBalanceHex);
+        });
+
+        it("should fund an account and permit that account to send a transaction", async function () {
+          // Arrange: Fund a not-yet-existing account.
+          const notYetExistingAccount =
+            "0x1234567890123456789012345678901234567890";
+          const amountToBeSent = new BN(10);
+          const gasRequired = new BN("48000000000000000");
```

Comment:
Is there a constant that could be used here instead?

Reply:
Is this the gas limit or the gas price? It seems a huge value to be the gas limit. If it's the gas price, you can probably omit it.

Reply:
Honestly I'm not sure.  I only stuck it in here because, without it, I get this error message: `InvalidInputError: sender doesn't have enough funds to send tx. The upfront cost is: 48000000000000010 and the sender's account only has: 10`.

Reply:
Ok, I understand this better now.

I think using an arbitrary big number like these is fine. But if you want to be more in control of what's going on, you can do this:

1) Use 21000 as the `gasRequired` here
2) When sending the transaction, set the `gas` value to 21000 and the `gasPrice` to 1
3) If you do this, then the resulting balance of `notYetExistingAccount` should be 0 again. You can assert that too if you want.